src/server/stream_family.cc

```diff
@@ -1402,44 +1533,131 @@ std::optional<ReadOpts> ParseReadArgsOrReply(CmdArgList args, ConnectionContext*
     return std::nullopt;
   }
 
+  vector<unordered_map<string_view, StreamIDsItem>> res_sitems(shard_set->size());
+  bool stop = false;
+  auto g_cb = [&](Transaction* t, EngineShard* shard) {
+    if (!opts.read_group) {
+      return OpStatus::OK;
+    }
+    auto sid = shard->shard_id();
+    auto s_args = t->GetShardArgs(sid);
+    unordered_map<string_view, StreamIDsItem> sid_items;
+
+    // get group and consumer
+    for (size_t i = 0; i < s_args.size(); i++) {
+      string_view key = s_args[i];
+
+      streamCG* group = nullptr;
+      streamConsumer* consumer = nullptr;
+      auto group_res = FindGroup(t->GetOpArgs(shard), key, opts.gname);
+      if (!group_res) {
+        (*cntx)->SendError(group_res.status());
+        stop = true;
+        return group_res.status();
+      }
+      if (group = group_res->second; !group) {
+        (*cntx)->SendError(NoGroupError(key, opts.gname));
+        stop = true;
+        return OpStatus::SKIPPED;
+      }
+
+      shard->tmp_str1 = sdscpylen(shard->tmp_str1, opts.cname.data(), opts.cname.size());
+      consumer = streamLookupConsumer(group, shard->tmp_str1, SLC_NO_REFRESH);
+      if (!consumer) {
+        consumer =
+            streamCreateConsumer(group, shard->tmp_str1, NULL, 0, SCC_NO_NOTIFY | SCC_NO_DIRTIFY);
+      }
+      StreamIDsItem sitem;
+      sitem.group = group;
+      sitem.consumer = consumer;
+      auto [_, is_inserted] = sid_items.emplace(key, sitem);
+      if (!is_inserted) {
+        (*cntx)->SendError("Same stream specified multiple times");
+        stop = true;
+        return OpStatus::SKIPPED;
+      }
+    }
+    res_sitems[sid] = sid_items;
+    return OpStatus::OK;
+  };
+
+  cntx->transaction->Schedule();
+  cntx->transaction->Execute(std::move(g_cb), false);
+
+  if (stop) {
+    goto close_transaction;
+  }
```

Comment:
goto is really bug prone. I will check it out explicitly and see if we can have a quick workaround and come back to you here. Otherwise, let's check thoroughly that things don't go wrong and there are no leaks whatsoever (I spotted 5 uses of `goto` and admittedly I am scared)

Reply:
@dranikpg do you know which PR?

Reply:
True, I also avoid using `goto`s. But didn't find any better way to do this

Reply:
It's #1425, its was on hold because it reduced the likelihood of catching blocking bugs

Reply:
I suggest that you duplicate the code here, and just run the callback and return early.